# SRC1000: Avoid System.Runtime.Caching.MemoryCache in .NET Core+

## Summary

The `System.Runtime.Caching.MemoryCache` class should not be used in .NET Core or .NET 5+ projects. Use `Microsoft.Extensions.Caching.Memory` instead.

## Rule Description

The `System.Runtime.Caching.MemoryCache` class was originally designed for .NET Framework and is included in .NET Core and .NET 5+ primarily as a compatibility bridge for porting applications. For modern .NET Core and .NET 5+ applications, Microsoft recommends using the `Microsoft.Extensions.Caching.Memory` namespace instead, as it provides better integration with the .NET Core dependency injection system and other platform services.

This analyzer reports a warning when `System.Runtime.Caching.MemoryCache` is used in .NET Core or .NET 5+ projects.

## How to Fix Violations

Replace uses of `System.Runtime.Caching.MemoryCache` with `Microsoft.Extensions.Caching.Memory.MemoryCache`.

### Before

```csharp
using System.Runtime.Caching;

class Program
{
    void CacheExample()
    {
        var cache = new MemoryCache("myCache");
        cache.Add("key", "value", DateTimeOffset.Now.AddMinutes(10));
        var value = cache.Get("key") as string;
    }
}
```

### After

```csharp
using Microsoft.Extensions.Caching.Memory;

class Program
{
    void CacheExample()
    {
        var options = new MemoryCacheOptions();
        var cache = new MemoryCache(options);
        
        // Note the API differences
        cache.Set("key", "value", TimeSpan.FromMinutes(10));
        var value = cache.Get<string>("key");
    }
}
```

### Microsoft.Extensions.Caching.Memory Integration with Dependency Injection

In ASP.NET Core or other services using dependency injection:

```csharp
// In Startup.cs or Program.cs
services.AddMemoryCache();

// In a service class
public class MyService
{
    private readonly IMemoryCache _cache;

    public MyService(IMemoryCache cache)
    {
        _cache = cache;
    }

    public void CacheExample()
    {
        _cache.Set("key", "value", TimeSpan.FromMinutes(10));
        var value = _cache.Get<string>("key");
    }
}
```

## When to Suppress

You might suppress this warning in the following scenarios:

1. During incremental migration from .NET Framework to .NET Core/5+
2. When maintaining a codebase that needs to work across both .NET Framework and .NET Core
3. When dependent libraries require System.Runtime.Caching and refactoring is not practical

To suppress:

```csharp
#pragma warning disable SRC1000 // Avoid System.Runtime.Caching.MemoryCache in .NET Core+
var cache = new MemoryCache("myCache");
#pragma warning restore SRC1000
```

## Example

### Noncompliant Code

```csharp
using System.Runtime.Caching;

class Example
{
    public void UseCache()
    {
        // Creates a diagnostic: SRC1000
        var cache = new MemoryCache("myCache");
        
        // Also creates diagnostics: SRC1000
        object value = cache.Get("myKey");
        cache.Set("myKey", "value", new CacheItemPolicy());
    }
}
```

### Compliant Code

```csharp
using Microsoft.Extensions.Caching.Memory;

class Example
{
    public void UseCache()
    {
        // No diagnostic - uses recommended API
        var cache = new MemoryCache(new MemoryCacheOptions());
        var value = cache.Get<object>("myKey");
        cache.Set("myKey", "value", TimeSpan.FromMinutes(10));
    }
}
```

## Related Resources

- [Microsoft.Extensions.Caching.Memory Documentation](https://learn.microsoft.com/dotnet/core/extensions/caching)
- [Microsoft.Extensions.Caching.Memory on NuGet](https://www.nuget.org/packages/Microsoft.Extensions.Caching.Memory)